### ISSUE TITLE
Feature request: shflags semantics

### ISSUE THREAD

**Issue by yetamrra** [koalaman/shellcheck#1597]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

[shflags](https://github.com/kward/shflags) is a popular library to process command-line flags (6000+ hits on github alone, plus hundreds more in Chromium OS).  It's not using any non-standard shell features, but scripts that use it produce a lot of shellcheck warnings because it creates variables at runtime and shellcheck doesn't know that's happening.  It also creates a few global variables that get warned about.

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

. "/path/to/shflags"

DEFINE_string my_flag "def" "Help text"
DEFINE_int unused_flag 0 "Help text 2"
DEFINE_boolean bool_flag ${FLAGS_TRUE} "More help text"

FLAGS_HELP="Usage message goes here."

FLAGS "$@" || exit $?
eval set -- "${FLAGS_ARGV}"

echo "${FLAGS_my_flag}"
echo "${FLAGS_no_such_flag}"

[[ ${FLAGS_bool_flag} -eq ${FLAGS_FALSE} ]] || echo "Bool set"
```

#### Here's what shellcheck currently says:
```
In flags.sh line 7:
DEFINE_boolean bool_flag ${FLAGS_TRUE} "More help text"
                         ^-----------^ SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
DEFINE_boolean bool_flag "${FLAGS_TRUE}" "More help text"


In flags.sh line 9:
FLAGS_HELP="Usage message goes here."
^--------^ SC2034: FLAGS_HELP appears unused. Verify use (or export if used externally).


In flags.sh line 14:
echo "${FLAGS_my_flag}"
      ^--------------^ SC2154: FLAGS_my_flag is referenced but not assigned.


In flags.sh line 15:
echo "${FLAGS_no_such_flag}"
      ^-------------------^ SC2154: FLAGS_no_such_flag is referenced but not assigned.


In flags.sh line 17:
[[ ${FLAGS_bool_flag} -eq ${FLAGS_FALSE} ]] || echo "Bool set"
   ^----------------^ SC2154: FLAGS_bool_flag is referenced but not assigned.
```

This has three shortcomings:

1. It misses that unused_flag isn't used anywhere.
2. It complains about several undefined FLAGS_ variables that are actually defined at runtime.
3. It suggests putting quotes around $FLAGS_TRUE even though it will never contain spaces.

#### Something like this would be more useful:

```
In flags.sh line 6:
DEFINE_int unused_flag 0 "Help text 2"
           ^---------^ SC2034: FLAGS_unused_flag appears unused. Verify use (or export if used externally).


In flags.sh line 15:
echo "${FLAGS_no_such_flag}"
      ^-------------------^ SC2154: FLAGS_no_such_flag is referenced but not assigned.
```

Adding the predefined shflags globals and treating DEFINE_* as syntax for creating FLAGS_ variables is pretty simple, but it's not totally clear how the behavior should be enabled.  A couple of ideas:

* A command-line flag or file-level directive to enable shflags semantics.
* Auto-enable if a source directive looks like it's requesting shflags.
* Always enable and just trust that DEFINE_xxx and FLAGS_xxx aren't likely to be used by anybody else (this one seems dubious).

Would any of those be palatable, or do you have other suggestions for adding this?

**Comment by koalaman** [comment]
Thanks for fixing this!
